### ISSUE TITLE
proper cmd for uploading/downloading preset from ah+fx

### ISSUE THREAD

**Issue by ianhundere** [dagargo/elektroid#128]
just curious what the appropriate command would be to upload/download presets from an ah+fx.

**Comment by dagargo** [comment]
First, I've just pushed a commit to fix an issue with the info command and updated the documentation too.

Probably you already know this but you can check which connectors and filesystems are available.

```
$ elektroid-cli ld
0: id: SYSTEM_ID; name: computer
1: id: hw:2,0,0; name: hw:2,0,0: Elektron Digitakt, Elektron Digitakt MIDI 1

$ elektroid-cli info 1
Type: MIDI
Device name: Elektron Digitakt
Device version: 1.51A
Device description: Digitakt
Connector name: elektron
Filesystems: sample, data (CLI only), project, sound
```

You should see this line for the Analog Heat filesystems.

```
Filesystems: data (CLI only), sound
```

From here, the operation commands are `ls`, `ul` and `dl` (list, upload and download). So these full commands should work for you.

```
$ elektroid-cli elektron-sound-ls 1:/
$ elektroid-cli elektron-sound-dl 1:/A/1
$ elektroid-cli elektron-sound-ul EXAMPLE.ahsnd 1:/B/1
```

Is everything working as expected?

Probably we should "rename" the filesystem from `sound` to `preset` to mimic the devices. Let me know your thoughts on this.

**Comment by ianhundere** [comment]
i pulled latest

before i was using `elektron-preset-ul` w/ the following output:
```
❯ elektroid-cli elektron-preset-ul file.ahpst 18:/presets
ERROR:connectors/package.c:616:(package_send_pkg_resources): Error while uploading payload to '/presets/BOOST COMP.mc-snd'
ERROR:elektroid-cli.c:896:(main): Error: Input/output error
```

here's other output i'm getting now w/ what you suggested:
```
❯ elektroid-cli elektron-sound-ls 18:/
ERROR:elektroid-cli.c:896:(main): Error: Not a directory
```
```
❯ elektroid-cli info 18
Type: MIDI
Device name: Elektron Analog Heat +FX
Device version: 1.01
Device description: Analog Heat +FX
Connector name: elektron
Filesystems: sample (CLI only), raw (CLI only), preset (CLI only), data, project (CLI only), sound
```

tho data gave me things when i was playing before:
```
❯ elektroid-cli elektron-data-ls 18:/presets/
F 007e 1 0       304B 001 MAGNETIC DREAMS
F 007e 1 0       304B 002 BOUNCER UP
F 007e 1 0       304B 003 GOOD COPY
F 007e 1 0       304B 004 BAD COPY
F 007e 1 0       304B 005 EERY ECHO
F 007e 1 0       304B 006 UNDER THE CARPET
F 007e 1 0       304B 007 RYMDKLANG
F 007e 1 0       304B 008 CORRUPTED SPACE
F 007e 1 0       304B 009 POMPEI BY NIGHT
F 007e 1 0       304B 010 SHRINKING ROOM
F 007e 1 0       304B 011 FILTER CRUNCHER
F 007e 1 0       304B 012 DRUM VICE
F 007e 1 0       304B 013 FOLLOW THE BUZZ
F 007e 1 0       304B 014 CARTRIDGE OXIDE
F 007e 1 0       304B 015 TYPE IV C60
F 007e 1 0       304B 016 ERODING NOSTALGIA
F 007e 1 0       304B 017 DANCING HAZE
F 007e 1 0       304B 018 SLIPPERY NOODLES
F 007e 1 0       304B 019 ANTI-CLIMAX
F 007e 1 0       304B 020 FIRST
F 007e 1 0       304B 021 FANCIFUL TOPSOIL
F 007e 1 0       304B 022 HEAVY COMP
F 007e 1 0       304B 023 ---
F 007e 1 0       304B 033 HIGHS SWEETENER
```

**Comment by ianhundere** [comment]
> I'll create a branch with the aforementioned changes so we could work from there and let you know.

sounds good, and as a followup the following are the extensions that show when connected to the transfer app:
<img width="236" alt="image" src="https://github.com/dagargo/elektroid/assets/138915/9c85c838-d0cb-4800-9059-3f00ad7b59cc">

**Comment by dagargo** [comment]
Code pushed to branch `feature_add_ah_preset_fs`.

A few things to consider.

* There are 3 types of filesystems implemented in the Elektron devices: samples, raw (used only on Model:Cycles AFAIK) and data.
* The data filesystem found in Elektroid provides direct access to what the devices provide. This is hidden in the GUI as it is  not compatible with the Transfer format and it is only meant to be used for testing purposes. Let's say that the data format is binary. It is not SysEx either.
* On top of the data filesytem, there are sounds, projects and presets (I was not aware of presets until today). These are in different data directories like the ones you were seeing but are actually part of the same filesystem.
* For these data filesystems, both Elektroid and Transfer add use a zip file containing a `manifest.json` file describing the content of the zip file and the binary file (Digitakt also includes the related samples). Therefore **this zip files are neither SysEx files nor the binary data files**. File extension should be `ahfxpst`.
* Regarding the last point, take care with your backups, please. You should be able to restore anything using the `data` filesystem as you were doing before but those files won't work the new preset filesystem.

Feel free to ask anything regarding these if it's not clear enough.

For reference, the content of the file should look like this. (This particular file is the manifest.json in the `BOOST COMP` sound included in the 10340° Sound Pack from Elektron.

```
{
  "FormatVersion": "1.0",
  "ProductType": [
    "10",
    "22"
  ],
  "Payload": "BOOST COMP",
  "FileType": "Sound",
  "FirmwareVersion": "DEV BUILD 0031"
}
```

Everything should work from the GUI. From the CLI, it won't as the Model:Cycles filesystem is called `preset` too and you will not be able to address the AH preset filesystem by its name. If you want to try the CLI, delete the line 3302.

https://github.com/dagargo/elektroid/blob/1a496b70b32c555fe2958707739d627161221d50/src/connectors/elektron.c#L3301-L3304

I'll fix this once the rest is confirmed to work.

Let me know if this fixes your issues.

**Comment by ianhundere** [comment]
so it all worked via the gui except that i had to drag from my file manager over to the ah+fx as nothing came up in the source folder(s):
![image](https://github.com/dagargo/elektroid/assets/138915/fc86b43b-5f1b-423a-9c95-a9768932fff9)

other than that, all good. nice work!

**Comment by dagargo** [comment]
That's great! Thank you.

Is it possible that you had no files with the `ahfxpst` extension on the left side (local browser)?

**Comment by ianhundere** [comment]
ah, i was looking at files w/ `ahpst` which is also a valid extension.

**Comment by dagargo** [comment]
Great! Everything is working as intended.

Regarding using more than one extension to filter files, this is currently not implemented. I'll take a look at it in the following days.

**Comment by dagargo** [comment]
I've just create #129 to add more than one extension when filtering files on the left side.

In the meantime, I've just pushed a commit addressing the CLI issues. Could you give it a try?
I've run my integration tests and everything seems to work but better if you could verify this too.

**Comment by dagargo** [comment]
I've done some more testing and I think everything is OK.
Code has merged in #130.

**Comment by ianhundere** [comment]
> I've done some more testing and I think everything is OK.
> Code has merged in #130.

ah sorry, was going to test this weekend.

**Comment by dagargo** [comment]
Sorry, I needed to push a few commits to keep on working on other features.
Thank you anyway. :pray:
